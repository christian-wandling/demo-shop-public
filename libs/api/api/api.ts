export * from './cart-items-api';
import { CartItemsApi } from './cart-items-api';
export * from './orders-api';
import { OrdersApi } from './orders-api';
export * from './products-api';
import { ProductsApi } from './products-api';
export * from './shopping-sessions-api';
import { ShoppingSessionsApi } from './shopping-sessions-api';
export * from './users-api';
import { UsersApi } from './users-api';
export const APIS = [CartItemsApi, OrdersApi, ProductsApi, ShoppingSessionsApi, UsersApi];
