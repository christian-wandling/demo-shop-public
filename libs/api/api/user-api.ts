/**
 * Demo Shop API
 *
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/* tslint:disable:no-unused-variable member-ordering */

import { Inject, Injectable, Optional } from '@angular/core';
import {
  HttpClient,
  HttpHeaders,
  HttpParams,
  HttpResponse,
  HttpEvent,
  HttpParameterCodec,
  HttpContext,
} from '@angular/common/http';
import { CustomHttpParameterCodec } from '../encoder';
import { Observable } from 'rxjs';

// @ts-ignore
import { AddressResponse } from '../model/address-response';
// @ts-ignore
import { UpdateUserAddressRequest } from '../model/update-user-address-request';
// @ts-ignore
import { UpdateUserPhoneRequest } from '../model/update-user-phone-request';
// @ts-ignore
import { UserResponse } from '../model/user-response';

// @ts-ignore
import { BASE_PATH, COLLECTION_FORMATS } from '../variables';
import { Configuration } from '../configuration';

@Injectable({
  providedIn: 'root',
})
export class UserApi {
  protected basePath = 'http://localhost';
  public defaultHeaders = new HttpHeaders();
  public configuration = new Configuration();
  public encoder: HttpParameterCodec;

  constructor(
    protected httpClient: HttpClient,
    @Optional() @Inject(BASE_PATH) basePath: string | string[],
    @Optional() configuration: Configuration
  ) {
    if (configuration) {
      this.configuration = configuration;
    }
    if (typeof this.configuration.basePath !== 'string') {
      const firstBasePath = Array.isArray(basePath) ? basePath[0] : undefined;
      if (firstBasePath != undefined) {
        basePath = firstBasePath;
      }

      if (typeof basePath !== 'string') {
        basePath = this.basePath;
      }
      this.configuration.basePath = basePath;
    }
    this.encoder = this.configuration.encoder || new CustomHttpParameterCodec();
  }

  // @ts-ignore
  private addToHttpParams(httpParams: HttpParams, value: any, key?: string): HttpParams {
    if (typeof value === 'object' && value instanceof Date === false) {
      httpParams = this.addToHttpParamsRecursive(httpParams, value);
    } else {
      httpParams = this.addToHttpParamsRecursive(httpParams, value, key);
    }
    return httpParams;
  }

  private addToHttpParamsRecursive(httpParams: HttpParams, value?: any, key?: string): HttpParams {
    if (value == null) {
      return httpParams;
    }

    if (typeof value === 'object') {
      if (Array.isArray(value)) {
        (value as any[]).forEach(elem => (httpParams = this.addToHttpParamsRecursive(httpParams, elem, key)));
      } else if (value instanceof Date) {
        if (key != null) {
          httpParams = httpParams.append(key, (value as Date).toISOString().substring(0, 10));
        } else {
          throw Error('key may not be null if value is Date');
        }
      } else {
        Object.keys(value).forEach(
          k => (httpParams = this.addToHttpParamsRecursive(httpParams, value[k], key != null ? `${key}.${k}` : k))
        );
      }
    } else if (key != null) {
      httpParams = httpParams.append(key, value);
    } else {
      throw Error('key may not be null if value is not object or array');
    }
    return httpParams;
  }

  /**
   * Resolve current user
   * Resolve current user based on identity extracted from bearer token
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public resolveCurrentUser(
    observe?: 'body',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<UserResponse>;
  public resolveCurrentUser(
    observe?: 'response',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpResponse<UserResponse>>;
  public resolveCurrentUser(
    observe?: 'events',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpEvent<UserResponse>>;
  public resolveCurrentUser(
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<any> {
    let localVarHeaders = this.defaultHeaders;

    let localVarCredential: string | undefined;
    // authentication (bearer) required
    localVarCredential = this.configuration.lookupCredential('bearer');
    if (localVarCredential) {
      localVarHeaders = localVarHeaders.set('Authorization', 'Bearer ' + localVarCredential);
    }

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['application/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/users/me`;
    return this.httpClient.request<UserResponse>('post', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Update address of current user
   * Update the address of the user based on identity extracted from bearer token
   * @param updateUserAddressRequest
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public updateCurrentUserAddress(
    updateUserAddressRequest: UpdateUserAddressRequest,
    observe?: 'body',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<AddressResponse>;
  public updateCurrentUserAddress(
    updateUserAddressRequest: UpdateUserAddressRequest,
    observe?: 'response',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpResponse<AddressResponse>>;
  public updateCurrentUserAddress(
    updateUserAddressRequest: UpdateUserAddressRequest,
    observe?: 'events',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpEvent<AddressResponse>>;
  public updateCurrentUserAddress(
    updateUserAddressRequest: UpdateUserAddressRequest,
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<any> {
    if (updateUserAddressRequest === null || updateUserAddressRequest === undefined) {
      throw new Error(
        'Required parameter updateUserAddressRequest was null or undefined when calling updateCurrentUserAddress.'
      );
    }

    let localVarHeaders = this.defaultHeaders;

    let localVarCredential: string | undefined;
    // authentication (bearer) required
    localVarCredential = this.configuration.lookupCredential('bearer');
    if (localVarCredential) {
      localVarHeaders = localVarHeaders.set('Authorization', 'Bearer ' + localVarCredential);
    }

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['application/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    // to determine the Content-Type header
    const consumes: string[] = ['application/json'];
    const httpContentTypeSelected: string | undefined = this.configuration.selectHeaderContentType(consumes);
    if (httpContentTypeSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Content-Type', httpContentTypeSelected);
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/users/me/address`;
    return this.httpClient.request<AddressResponse>('put', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      body: updateUserAddressRequest,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Update phone of current user
   * Update the phone of the user based on identity extracted from bearer token
   * @param updateUserPhoneRequest
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public updateCurrentUserPhone(
    updateUserPhoneRequest: UpdateUserPhoneRequest,
    observe?: 'body',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<UserResponse>;
  public updateCurrentUserPhone(
    updateUserPhoneRequest: UpdateUserPhoneRequest,
    observe?: 'response',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpResponse<UserResponse>>;
  public updateCurrentUserPhone(
    updateUserPhoneRequest: UpdateUserPhoneRequest,
    observe?: 'events',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpEvent<UserResponse>>;
  public updateCurrentUserPhone(
    updateUserPhoneRequest: UpdateUserPhoneRequest,
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<any> {
    if (updateUserPhoneRequest === null || updateUserPhoneRequest === undefined) {
      throw new Error(
        'Required parameter updateUserPhoneRequest was null or undefined when calling updateCurrentUserPhone.'
      );
    }

    let localVarHeaders = this.defaultHeaders;

    let localVarCredential: string | undefined;
    // authentication (bearer) required
    localVarCredential = this.configuration.lookupCredential('bearer');
    if (localVarCredential) {
      localVarHeaders = localVarHeaders.set('Authorization', 'Bearer ' + localVarCredential);
    }

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['application/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    // to determine the Content-Type header
    const consumes: string[] = ['application/json'];
    const httpContentTypeSelected: string | undefined = this.configuration.selectHeaderContentType(consumes);
    if (httpContentTypeSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Content-Type', httpContentTypeSelected);
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/users/me/phone`;
    return this.httpClient.request<UserResponse>('patch', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      body: updateUserPhoneRequest,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }
}
