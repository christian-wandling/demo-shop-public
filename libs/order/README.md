# `@demo-shop/order`

This library handles aspects of the order feature including order selection and invoice pdf generation.

## Features

- **Order Management**

  - List orders history of current user

- **Order Details**

  - Comprehensive order information display
  - Product details with images and pricing

- **Invoice Generation**

  - In-browser PDF generation and download

- **State Management**
  - Signal-based reactive state
  - Synchronization with backend

## API Reference

For a complete API reference, see the [Compodoc documentation](/apps/frontend/docs/compodoc).

## Running unit tests

Run `nx test order` to execute the unit tests.
