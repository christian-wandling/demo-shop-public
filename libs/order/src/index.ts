export * from './lib/order.routes';
export * from './lib/order.facade';
