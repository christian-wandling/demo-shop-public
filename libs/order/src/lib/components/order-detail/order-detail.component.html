@if (user() && order()) {
<div class="mx-auto mb-8 px-8 max-w-5xl" id="invoice">
  <section class="py-8 top-16 sticky z-10 border-b-2 border-gray-100 bg-white flex justify-between">
    <div class="flex flex-col gap-5">
      <div>
        <h1 class="text-s leading-tight font-bold text-gray-700">{{ user()?.firstname }} {{ user()?.lastname }}</h1>
        <div class="text-xs text-gray-500">
          <p>{{ user()?.address?.street }} {{ user()?.address?.apartment }}</p>
          <p>{{ user()?.address?.zip }} {{ user()?.address?.city }}, {{ user()?.address?.region }}</p>
          <p>{{ user()?.address?.country }}</p>
        </div>
      </div>
      <button
        (click)="printPdf(order(), user())"
        class="btn-print w-36 flex items-center justify-center rounded-md border border-transparent px-8 py-2 text-base font-medium bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2':">
        Print
      </button>
    </div>
    <div class="flex flex-col gap-2.5">
      <div class="flex flex-col items-end">
        <div class="text-s leading-tight font-semibold">Order #{{ order().id }}</div>
        <lib-date-time [dateTime]="order().created" class="text-xs text-gray-500" />
      </div>
      <lib-order-status [status]="order().status" />
    </div>
  </section>
  <section>
    <ul class="divide-y divide-gray-100">
      @for (item of order().items; track item.productId) {
      <li class="flex flex-col sm:flex-row gap-2 sm:gap-6 py-5">
        <div class="relative w-12 sm:w-24 h-12 sm:h-24 aspect-1 rounded-lg flex-shrink-0 overflow-hidden">
          <img [ngSrc]="item.productThumbnail" [alt]="item.productName" fill="" placeholder />
        </div>
        <div class="flex flex-col min-w-0 grow">
          <p class="text-sm sm:leading-6 text-gray-900">#{{ item.productId }} {{ item.productName }}</p>
          <p class="sm:mt-1 truncate text-xs sm:leading-5 text-gray-500">Qty {{ item.quantity }}</p>
        </div>
        <div class="flex flex-row sm:flex-col-reverse grow sm:grow-0 justify-between sm:justify-end sm:items-end">
          <p class="sm:mt-1 text-xs leading-5 text-gray-500">
            <span>{{ item.unitPrice | currency }}</span>
          </p>
          <p class="text-sm leading-6 text-gray-900">{{ item.totalPrice | currency }}</p>
        </div>
      </li>
      }
      <li class="flex justify-between gap-x-6 py-5">
        <div class="flex min-w-0 gap-x-4">
          <div class="min-w-0 flex-auto">
            <p class="text-sm font-semibold leading-6 text-gray-900">total</p>
          </div>
        </div>
        <div class="shrink-0 sm:flex sm:flex-col sm:items-end">
          <p class="text-sm leading-6 text-gray-900">{{ order().amount | currency }}</p>
        </div>
      </li>
    </ul>
  </section>
</div>
}
