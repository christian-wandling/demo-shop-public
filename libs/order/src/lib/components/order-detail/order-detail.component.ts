import { ChangeDetectionStrategy, Component, inject, OnInit } from '@angular/core';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { OrderFacade } from '../../order.facade';
import { ActivatedRoute } from '@angular/router';
import { UserFacade } from '@demo-shop/user';
import { PrintInvoiceService } from '../../services/print-invoice.service';
import { OrderResponse, UserResponse } from '@demo-shop/api';
import { OrderStatusComponent } from '../shared/order-status/order-status.component';
import { DateTimeComponent } from '@demo-shop/shared';

@Component({
  selector: 'lib-order-detail',
  standalone: true,
  imports: [CommonModule, NgOptimizedImage, OrderStatusComponent, DateTimeComponent],
  templateUrl: './order-detail.component.html',
  styleUrl: './order-detail.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class OrderDetailComponent implements OnInit {
  readonly #userFacade = inject(UserFacade);
  readonly user = this.#userFacade.getCurrentUser();
  readonly #orderFacade = inject(OrderFacade);
  readonly #activatedRoute = inject(ActivatedRoute);
  readonly order = this.#orderFacade.find(this.#activatedRoute.snapshot.params['id']);
  readonly #printInvoiceService = inject(PrintInvoiceService);

  ngOnInit(): void {
    this.#orderFacade.fetchById(this.#activatedRoute.snapshot.params['id']);
  }

  printPdf(order: OrderResponse, user: UserResponse | undefined): void {
    if (!user) {
      console.error('Could not find user');
      return;
    }

    this.#printInvoiceService.generatePdf(order, user);
  }
}
