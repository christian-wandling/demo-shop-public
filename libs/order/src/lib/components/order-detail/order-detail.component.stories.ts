import { applicationConfig, Meta, StoryObj } from '@storybook/angular';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { ActivatedRoute } from '@angular/router';
import { importProvidersFrom, signal } from '@angular/core';
import { AddressResponse, OrderItemResponse, OrderResponse, OrderStatus, UserResponse } from '@demo-shop/api';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { faker } from '@faker-js/faker';
import { provideImageLoader } from '@demo-shop/shared';
import { OrderDetailComponent } from './order-detail.component';
import { OrderFacade } from '../../order.facade';
import { UserFacade } from '@demo-shop/user';
import { PrintInvoiceService } from '../../services/print-invoice.service';

const items: OrderItemResponse[] = Array.from({ length: 5 }).map((_, productId) => {
  const quantity = faker.number.int({ min: 1, max: 5 });
  const unitPrice = faker.number.int({ min: 1, max: 20 }) * 10 - 1;

  return {
    productId,
    productName: faker.commerce.productName(),
    productThumbnail: faker.image.urlPicsumPhotos(),
    quantity,
    unitPrice,
    totalPrice: quantity * unitPrice,
  };
});

const order: OrderResponse = {
  id: 1,
  amount: items.reduce((acc, curr) => acc + curr.totalPrice, 0),
  created: faker.date.past().toISOString(),
  items,
  status: OrderStatus.Created,
  userId: 0,
};

const address: AddressResponse = {
  street: faker.location.streetAddress(),
  apartment: faker.number.int({ max: 100 }).toString(),
  city: faker.location.city(),
  zip: faker.location.zipCode('#####'),
  region: faker.location.state(),
  country: faker.location.country(),
};

const user: UserResponse = {
  id: 1,
  email: faker.internet.email(),
  firstname: faker.person.firstName(),
  lastname: faker.person.lastName(),
  phone: faker.helpers.fromRegExp('+[0-9]{9}'),
  address,
};

const mockOrderFacade = {
  getById: () => signal(order),
  fetchById: () => {
    return;
  },
};

const mockUserFacade = {
  getCurrentUser: () => signal(user),
};

const mockActivatedRoute = {
  snapshot: {
    params: {
      id: 1,
    },
  },
};

const mockPrintInvoiceService = {
  generatePdf: () => {
    return;
  },
};

const meta: Meta<OrderDetailComponent> = {
  component: OrderDetailComponent,
  title: 'Order/OrderDetailComponent',
  decorators: [
    applicationConfig({
      providers: [
        importProvidersFrom(CommonModule, NgOptimizedImage),
        provideImageLoader(),
        { provide: OrderFacade, useValue: mockOrderFacade },
        { provide: ActivatedRoute, useValue: mockActivatedRoute },
        { provide: UserFacade, useValue: mockUserFacade },
        { provide: PrintInvoiceService, useValue: mockPrintInvoiceService },
      ],
    }),
  ],
  argTypes: {
    printPdf: { action: 'printPdf', control: false },
  },
  parameters: {
    controls: {
      exclude: ['#activatedRoute', '#orderFacade', '#userFacade', '#printInvoiceService', 'order', 'user', 'ngOnInit'],
    },
  },
};
export default meta;
type Story = StoryObj<OrderDetailComponent>;

export const Default: Story = {
  render: args => ({
    props: args,
    template: `
      <div>
        <div style="height: 4rem; width: 100%; background: #fff; z-index: 10; top: 0; position: sticky"></div>
        <lib-order-detail></lib-order-detail>
      </div>
    `,
  }),
  play: async ({ canvasElement, args }) => {
    const name = within(canvasElement).getByText(`${user.firstname} ${user.lastname}`);
    const addressLine1 = within(canvasElement).getByText(`${address.street} ${address.apartment}`);
    const addressLine2 = within(canvasElement).getByText(`${address.zip} ${address.city}, ${address.region}`);
    const addressLine3 = within(canvasElement).getByText(`${address.country}`);
    const printPdfButton = within(canvasElement).getByRole('button');
    const productImages = within(canvasElement).getAllByRole('img');
    expect(name).toBeTruthy();
    expect(addressLine1).toBeTruthy();
    expect(addressLine2).toBeTruthy();
    expect(addressLine3).toBeTruthy();
    expect(printPdfButton).toBeTruthy();
    expect(productImages).toHaveLength(items.length);
  },
};
