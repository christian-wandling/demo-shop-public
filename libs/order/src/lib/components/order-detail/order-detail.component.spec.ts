import { ComponentFixture, TestBed } from '@angular/core/testing';
import { OrderDetailComponent } from './order-detail.component';
import { UserFacade } from '@demo-shop/user';
import { OrderDTO, OrderStatus } from '@demo-shop/api';
import { signal } from '@angular/core';
import { OrderFacade } from '../../order.facade';
import { ActivatedRoute } from '@angular/router';
import { PrintInvoiceService } from '../../services/print-invoice.service';
import { UTCDate } from '@date-fns/utc';

describe('OrderDetailComponent', () => {
  let component: OrderDetailComponent;
  let fixture: ComponentFixture<OrderDetailComponent>;
  let printInvoiceService: PrintInvoiceService;

  const mockUser = {
    address: {
      street: 'street',
      apartment: 'apartment',
      city: 'city',
      region: 'region',
      zip: 'zip',
      country: 'country',
    },
    email: 'email',
    firstname: 'firstname',
    id: 'id',
    lastname: 'lastname',
    phone: 'phone',
  };

  const mockOrder: OrderDTO = {
    id: '1',
    userId: '1',
    items: [],
    amount: 0,
    status: OrderStatus.Created,
    created: new UTCDate(2020, 0, 1).toString(),
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [OrderDetailComponent],
      providers: [
        { provide: UserFacade, useValue: { getCurrentUser: jest.fn().mockReturnValue(signal(mockUser)) } },
        { provide: OrderFacade, useValue: { find: jest.fn().mockReturnValue(signal(mockOrder)) } },
        {
          provide: ActivatedRoute,
          useValue: {
            snapshot: {
              params: {
                id: 'id',
              },
            },
          },
        },
        { provide: PrintInvoiceService, useValue: { generatePdf: jest.fn() } },
      ],
    }).compileComponents();

    printInvoiceService = TestBed.inject(PrintInvoiceService);
    fixture = TestBed.createComponent(OrderDetailComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('printPdf', () => {
    it('should call generatePDF when user is defined', () => {
      component.printPdf(mockOrder, mockUser);

      expect(printInvoiceService.generatePdf).toHaveBeenCalledWith(mockOrder, mockUser);
    });

    it('should not call generatePDF when user is undefined', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementationOnce(() => undefined);

      component.printPdf(mockOrder, undefined);

      expect(printInvoiceService.generatePdf).not.toHaveBeenCalled();
      expect(consoleSpy).toHaveBeenCalledWith('Could not find user');

      consoleSpy.mockRestore();
    });
  });
});
