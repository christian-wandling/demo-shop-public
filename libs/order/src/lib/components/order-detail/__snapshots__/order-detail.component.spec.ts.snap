// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderDetailComponent should match the snapshot 1`] = `
<lib-order-detail
  order={[Function Function]}
  user={[Function Function]}
>
  <div
    class="mx-auto mb-8 px-8 max-w-5xl"
    id="invoice"
  >
    <section
      class="py-8 mb-2 flex justify-between items-end"
    >
      <div>
        <div
          class="flex flex-row gap-3 items-center justify-start"
        >
          <div
            class="leading-none"
          >
            <div
              class="text-xl leading-none font-semibold"
            >
              Order #1
            </div>
            <lib-date-time
              class="h-3 text-xs leading-none text-gray-500"
            >
              <time
                datetime="Wed Jan 1 2020 00:00:00 GMT+0000 (Coordinated Universal Time)"
              >
                Jan 01, 2020
              </time>
            </lib-date-time>
          </div>
          <lib-order-status
            class="rounded-lg px-4 py-1.5 max-w-28 text-center text-white border border-slate-300 text-xs bg-orange-500"
          >
            Created
          </lib-order-status>
        </div>
        <h1
          class="mt-5 text-s font-bold text-gray-700"
        >
          firstname lastname
        </h1>
        <div
          class="text-xs text-gray-500"
        >
          <p>
            street apartment
          </p>
          <p>
            zip city, region
          </p>
          <p>
            country
          </p>
        </div>
      </div>
      <button
        class="btn-print w-44 flex items-center justify-center rounded-md border border-transparent mb-3 px-8 py-2 text-base font-medium bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2':"
      >
         Print 
      </button>
    </section>
    <section>
      <ul
        class="divide-y divide-gray-100"
      >
        <li
          class="flex justify-between gap-x-6 py-5"
        >
          <div
            class="flex min-w-0 gap-x-4"
          >
            <div
              class="min-w-0 flex-auto"
            >
              <p
                class="text-sm font-semibold leading-6 text-gray-900"
              >
                total
              </p>
            </div>
          </div>
          <div
            class="hidden shrink-0 sm:flex sm:flex-col sm:items-end"
          >
            <p
              class="text-sm leading-6 text-gray-900"
            >
              $0.00
            </p>
          </div>
        </li>
      </ul>
    </section>
  </div>
</lib-order-detail>
`;
